### PR TITLE
Add currently-unused tlv module

### PR DESCRIPTION
**Pull request by jgallagher**

### CODE REVIEW

File: gateway-messages/src/tlv.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    let (header_buf, out) = out.split_at_mut(mem::size_of::<Header>());
+
+    // Generate the `value` directly into the buffer.
+    let value_len = value(out).map_err(EncodeError::Custom)?;
+
+    // `value` should never claim to have used more data than exists in `out`.
+    assert!(value_len <= out.len());
+
+    // Convert to u32 or die trying.
```
Comment (andrewjstone) [inline gateway-messages/src/tlv.rs:73]: lol

File: gateway-messages/src/tlv.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DecodeError {
+    /// The buffer is too small to contain a tag/length header.
+    BufferTooSmall,
+    /// The `length` field requires more data than is remaining in the buffer.
+    LengthTooLong,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, AsBytes, FromBytes)]
+#[repr(C)]
```
Comment (andrewjstone) [inline gateway-messages/src/tlv.rs:32]: Just curios, why the `repr(C)` here?

**Review by andrewjstone — APPROVED** [review]
Looks great. You made quick work here!

File: gateway-messages/src/tlv.rs (hunk 2 above)
Comment (jgallagher) [reply gateway-messages/src/tlv.rs:32]: It's a requirement to derive the zerocopy traits: https://docs.rs/zerocopy/latest/zerocopy/trait.AsBytes.html#safety

**Pull request merged**